File: java/src/main/java/io/github/gsoultan/anubis/Permission.java
package io.github.gsoultan.anubis;

import java.util.Objects;

/**
 * A full permission key: {@code app:resource:action}.
 *
 * <p>The full key is what tokens carry and what {@code authorize} takes. Inside
 * an application manifest the same permission is written WITHOUT the
 * application prefix — {@code invoice:approve} — and that asymmetry has cost
 * people real time. {@link #manifest()} makes the two forms convertible instead
 * of a thing to remember.
 */
public final class Permission {
    private final String key;
    private final String[] parts; // null when malformed

    private Permission(String key) {
        this.key = key == null ? "" : key;
        String[] segments = this.key.split(":", -1);
        boolean wellFormed = segments.length == 3;
        if (wellFormed) {
            for (String segment : segments) {
                if (segment.isEmpty()) {
                    wellFormed = false;
                    break;
                }
            }
        }
        this.parts = wellFormed ? segments : null;
    }

    public static Permission of(String key) {
        return new Permission(key);
    }

    public static Permission from(String app, String resource, String action) {
        return new Permission(app + ":" + resource + ":" + action);
    }

    public String key() {
        return key;
    }

    /** The application slug the permission belongs to, or "" when malformed. */
    public String app() {
        return parts == null ? "" : parts[0];
    }

    public String resource() {
        return parts == null ? "" : parts[1];
    }

    public String action() {
        return parts == null ? "" : parts[2];
    }

    /** Whether the key has the three non-empty parts Anubis requires. */
    public boolean isValid() {
        return parts != null;
    }

    /** The permission as a manifest writes it: {@code resource:action}. */
    public String manifest() {
        return parts == null ? key : parts[1] + ":" + parts[2];
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Permission other && key.equals(other.key);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(key);
    }

    @Override
    public String toString() {
        return key;
    }
}

File: php/src/Permission.php
<?php

declare(strict_types=1);

namespace Anubis;

/**
 * A full permission key: "app:resource:action".
 *
 * The full key is what tokens carry and what authorize() takes. Inside an
 * application manifest the same permission is written WITHOUT the application
 * prefix — "invoice:approve" — and that asymmetry has cost people real time.
 * manifest() makes the two forms convertible instead of a thing to remember.
 */
final readonly class Permission implements \Stringable
{
    /** @var list<string>|null the three parts, or null when malformed */
    private ?array $parts;

    private function __construct(public string $key)
    {
        $segments = explode(':', $key);
        $this->parts = count($segments) === 3 && !in_array('', $segments, true) ? $segments : null;
    }

    public static function of(self|string $key): self
    {
        return $key instanceof self ? $key : new self($key);
    }

    public static function from(string $app, string $resource, string $action): self
    {
        return new self("{$app}:{$resource}:{$action}");
    }

    /** The application slug the permission belongs to, or "" when malformed. */
    public function app(): string
    {
        return $this->parts[0] ?? '';
    }

    public function resource(): string
    {
        return $this->parts[1] ?? '';
    }

    public function action(): string
    {
        return $this->parts[2] ?? '';
    }

    /** Whether the key has the three non-empty parts Anubis requires. */
    public function isValid(): bool
    {
        return $this->parts !== null;
    }

    /** The permission as a manifest writes it: "resource:action", no prefix. */
    public function manifest(): string
    {
        return $this->parts === null ? $this->key : "{$this->parts[1]}:{$this->parts[2]}";
    }

    public function equals(self|string $other): bool
    {
        return $this->key === self::of($other)->key;
    }

    public function __toString(): string
    {
        return $this->key;
    }
}
